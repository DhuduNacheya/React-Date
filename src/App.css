.App {
  width: 30%;
  height: 300px;
  background-color: #fff;
  position: absolute;
  top:0;
  bottom: 0;
  left: 0;
  right: 0;
  margin: auto;
  border: #ddd 1px solid;
  box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
}

.header-cls{
height: 50px;
line-height: 50px;
 background-color:#009688!important;
 margin-bottom: 15px;
 font-weight: bold;
 color: white;
 box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
 padding-left:10px;
    

}


.m-bl-10{
  margin: 0px 10px 10px 10px;
  
}

.algn{
  text-align: center;
}

@media only screen and (max-width: 900px) {
  .App {
    width: 33%;
    min-width:200px;
    height: 400px;
    background-color: #fff;
    position: absolute;
    top:0;
    bottom: 0;
    left: 0;
    right: 0;
    margin: auto;
    border: #ddd 1px solid;
    box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
  }
  
  .header-cls{
  height: 50px;
  line-height: 28px;
   background-color:#009688!important;
   margin-bottom: 10px;
   font-weight: bold;
   color: white;
   box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
   padding-left:10px;
      
  
  }
}


@media (max-width:900px) and (min-width:729px) {
  .App {
    width: 33%;
    min-width:200px;
    height: 380px;
    background-color: #fff;
    position: absolute;
    top:0;
    bottom: 0;
    left: 0;
    right: 0;
    margin: auto;
    border: #ddd 1px solid;
    box-shadow: 0 4px 8px 0 rgba(0,0,0,0.2);
  }
  
}