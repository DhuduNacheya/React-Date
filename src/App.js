import React, { Component } from 'react';
import './App.css';
import { Button,Alert} from 'reactstrap';

import InputDate from './component/input';

class App extends Component {
  constructor(props){
      super(props);
      this.state={
          startDate:null,
          startDateText:"From Date",
          startDateValid:true,
          endDate:null,
          endDateText:"To Date",
          endDateValid:true,
          disabled:true,
          message:"Please select to and from dates.",
          finalResult:null,
          alert:"info"
         }
      this.dateStateHandler=this.dateStateHandler.bind(this);
      this.calculateHandler=this.calculateHandler.bind(this);
  }


  dateStateHandler(date,id){
   
   
   if(id==="startDate"){
        
        if(!date){

          if(!this.state.endDate){
            this.setState({disabled:true,
                        startDateValid:false,
                        endDateValid:false,
                        message:"Please select to and from dates.",
                        startDate:date,
                        finalResult:null,
                        alert:"danger"});
                        }
          
          else{
          this.setState({disabled:true,
                        startDateValid:false,
                        message:"Please select a from date.",
                        startDate:date,
                        finalResult:null,
                        alert:"danger"});
          }
                  }
        else{
              if(!this.state.endDate){
                this.setState({disabled:true,
                              startDate:date,
                              endDateValid:false,
                              startDateValid:true,
                              message:"Please select a to date.",
                              finalResult:null,
                              alert:"danger"});
              }
              else if(new Date(date)>new Date(this.state.endDate)){
                this.setState({disabled:true,
                            startDateValid:false,
                            startDate:date,
                            message:"To date must be after from date.",
                            finalResult:null,
                            alert:"danger"});
              }
              else{
                this.setState({disabled:false,
                                startDateValid:true,
                                startDate:date,
                                endDateValid:true,
                                message:"Please click on calculate to get days count.",
                                finalResult:null,
                                alert:"info"});
           
                 }
            }

       }
   else{
     if(!date){
         if(!this.state.startDate){
           this.setState({disabled:true,
                          startDateValid:false,
                          endDateValid:false,
                          message:"Please select to and from dates.",
                          endDate:date,
                          finalResult:null,
                          alert:"danger"});
          }
      else{
           this.setState({disabled:true,
                          endDate:date,
                          finalResult:null,
                          alert:"danger",
                          endDateValid:false,
                          message:"Please select a to date."});
          }     
    }
    else{
          if(!this.state.startDate){
            this.setState({disabled:true,
                          finalResult:null,
                          alert:"danger",
                          endDate:date,
                          startDateValid:false,
                          endDateValid:true,
                          message:"Please select a from date."});
          }
          else if(new Date(date)<new Date(this.state.startDate)){
            this.setState({disabled:true,
                            endDateValid:false,
                            finalResult:null,
                            endDate:date,
                            alert:"danger",
                            message:"To date must be after from date."});
          }
          else{
            this.setState({disabled:false,
                            startDateValid:true,
                            endDateValid:true,
                            endDate:date,
                            finalResult:null,
                            message:"Please click on calculate to get days count.",
                            alert:"info"});
           
          }
        }
     }  
  }

  calculateHandler(){           
   let t1 = new Date(this.state.startDate);
   let t2 = new Date(this.state.endDate);
   let result=Math.floor((Date.UTC(t2.getFullYear(), t2.getMonth(), t2.getDate()) - Date.UTC(t1.getFullYear(), t1.getMonth(), t1.getDate()) ) /(1000 * 60 * 60 * 24));
    result="Number of Days between "+t1.getDate()+"/"+(t1.getMonth()+1)+"/"+t1.getFullYear()+" and "
            +t2.getDate()+"/"+(t2.getMonth()+1) +"/"+t2.getFullYear()+" are "+result+ " Days.";
   
   this.setState({finalResult:result,
                   message:null}); 
    
  }


  render() {
    return (
      <div className={"App"}>
        
        <div className={"header-cls"}>{"Day Difference Calculator"}</div>

         <div className={"m-bl-10"}>
               {!this.state.message?null:<Alert color={this.state.alert}> {this.state.message}</Alert>}
        </div>

        <div className={"m-bl-10"}>
           
            <InputDate valid={this.state.startDateValid} dateText={this.state.startDateText} date={this.state.startDate} chng={(e)=>this.dateStateHandler(e,"startDate")}/>
        </div>
         
        <div className={"m-bl-10"}>
            
             <InputDate valid={this.state.endDateValid} dateText={this.state.endDateText} date={this.state.endDate} chng={(e)=>this.dateStateHandler(e,"endDate")}/>
        </div>


        <div className={"m-bl-10 algn"}>
            <Button disabled={this.state.disabled} color={this.state.disabled?"secondary":"primary"} onClick={this.calculateHandler}>Calculate</Button>
        </div>  
      
        <div className={"m-bl-10"}>
        {!this.state.finalResult?null:<Alert color="info">{this.state.finalResult}</Alert>}
        </div>
    
       </div>
    );
  }
}

export default App;
