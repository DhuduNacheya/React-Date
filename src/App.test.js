/* global describe, it, test, expect,jest */
import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import InputDate from './component/input';
import Enzyme,{ shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});


describe("App", () => {
  const wrapper = shallow(<App />);
  test("should render app", () => {
    expect(wrapper.getElements()).toMatchSnapshot();
  });
  
  test('start Date', () => {
  expect(wrapper.state().startDate).toEqual(null);
  });

  test("startDateText", () => {
    expect(wrapper.state().startDateText).toEqual("From Date");
  });

  test('end Date', () => {
     expect(wrapper.state().endDate).toEqual(null);
   });

  test('startDateValid', () => {
    expect(wrapper.state().startDateValid).toEqual(true);
    });
  
  test('endDateValid', () => {
      expect(wrapper.state().endDateValid).toEqual(true);
     });

  test('disabled', () => {
      expect(wrapper.state().disabled).toEqual(true);
    });

  test('message', () => {
      expect(wrapper.state().message).toEqual("Please select to and from dates.");
     });

  test('finalResult', () => {
      expect(wrapper.state().finalResult).toEqual(null);
      });

  test('alert', () => {
       expect(wrapper.state().alert).toEqual("info");
         });
      

   test("always renders a div", () => {
          const divs = wrapper.find(".App");
          expect(divs.length).toBeGreaterThan(0);
        });  
    
        
   test("always renders a button", () => {
      const btn=wrapper.find('Button');
          expect(btn.length).toBe(1);
      });
      
  test("button handler click", () => {
    const clickHandler = jest.fn();
     const butn=shallow(<btn onClick={clickHandler}/>);
      butn.simulate('click');
      expect(clickHandler).toHaveBeenCalled();
    });
  
});

describe("Input Date", () => {
  const wrapper = shallow(<InputDate/>);
  test("should render Input date", () => {
    expect(wrapper.getElements()).toMatchSnapshot();
  });
  
  test(" Input date handler call", () => {
    const changeHandler = jest.fn();
    const pckr=wrapper.find('DatePicker');
    const pcWrpr=shallow(<pckr onChange={changeHandler}/>);
    pcWrpr.simulate('change');
    expect(changeHandler).toHaveBeenCalled();
  });

});
  
