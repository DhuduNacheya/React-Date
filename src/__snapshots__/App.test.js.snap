// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render app 1`] = `
Array [
  <div
    className="App"
  >
    <div
      className="header-cls"
    >
      Day Difference Calculator
    </div>
    <div
      className="m-bl-10"
    >
      <Alert
        closeAriaLabel="Close"
        color="info"
        fade={true}
        isOpen={true}
        tag="div"
        transition={
          Object {
            "appear": true,
            "baseClass": "fade",
            "baseClassActive": "show",
            "enter": true,
            "exit": true,
            "in": true,
            "mountOnEnter": false,
            "onEnter": [Function],
            "onEntered": [Function],
            "onEntering": [Function],
            "onExit": [Function],
            "onExited": [Function],
            "onExiting": [Function],
            "tag": "div",
            "timeout": 150,
            "unmountOnExit": true,
          }
        }
      >
         
        Please select to and from dates.
      </Alert>
    </div>
    <div
      className="m-bl-10"
    >
      <InputDate
        chng={[Function]}
        date={null}
        dateText="From Date"
        valid={true}
      />
    </div>
    <div
      className="m-bl-10"
    >
      <InputDate
        chng={[Function]}
        date={null}
        dateText="To Date"
        valid={true}
      />
    </div>
    <div
      className="m-bl-10 algn"
    >
      <Button
        color="secondary"
        disabled={true}
        onClick={[Function]}
        tag="button"
      >
        Calculate
      </Button>
    </div>
    <div
      className="m-bl-10"
    />
  </div>,
]
`;

exports[`Input Date should render Input date 1`] = `
Array [
  <div
    className="main"
  >
    <div
      className="child"
    >
      undefined :
    </div>
    <div
      className="child"
    >
        
      <DatePicker
        allowSameDay={false}
        className="error"
        dateFormat="dd/MM/YYYY"
        dateFormatCalendar="LLLL yyyy"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        monthsShown={1}
        nextMonthButtonLabel="Next month"
        onBlur={[Function]}
        onChange={[Function]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onInputClick={[Function]}
        onInputError={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        preventOpenOnFocus={false}
        previousMonthButtonLabel="Previous Month"
        readOnly={false}
        renderDayContents={[Function]}
        shouldCloseOnSelect={true}
        showTimeSelect={false}
        timeCaption="Time"
        timeIntervals={30}
        withPortal={false}
      />
    </div>
  </div>,
]
`;
